cli/azd/pkg/ext/hooks.go

```diff
@@ -0,0 +1,247 @@
+			matchingScripts = append(matchingScripts, scriptConfig)
+		}
+	}
+
+	return matchingScripts
+}
+
+func (h *CommandHooks) execScript(ctx context.Context, scriptConfig *ScriptConfig) error {
+	// Delete any temporary inline scripts after execution
+	defer func() {
+		if scriptConfig.Location == ScriptLocationInline {
+			os.Remove(scriptConfig.Path)
+		}
+	}()
+
+	script, err := h.GetScript(scriptConfig)
+	if err != nil {
+		return err
+	}
+
+	// When running in an interactive terminal broadcast a message to the dev to remind them that custom hooks are running.
+	if h.interactive {
+		h.console.Message(
+			ctx,
+			output.WithBold(
+				fmt.Sprintf(
+					"Executing %s command hook => %s",
+					output.WithHighLightFormat(scriptConfig.Name),
+					output.WithHighLightFormat(scriptConfig.Path),
+				),
+			),
+		)
+	}
+
+	log.Printf("Executing script '%s'", scriptConfig.Path)
+	_, err = script.Execute(ctx, scriptConfig.Path, h.interactive)
+	if err != nil {
+		return fmt.Errorf("failed executing script '%s' : %w", scriptConfig.Path, err)
+	}
+
+	return nil
+}
+
+// Gets the script to execute based on the script configuration values
+// For inline scripts this will also create a temporary script file to execute
+func (h *CommandHooks) GetScript(scriptConfig *ScriptConfig) (tools.Script, error) {
+	if scriptConfig.Location == "" {
+		if scriptConfig.Path != "" {
+			scriptConfig.Location = ScriptLocationPath
+		} else if scriptConfig.Script != "" {
+			scriptConfig.Location = ScriptLocationInline
+		}
+	}
+
+	if scriptConfig.Location == ScriptLocationInline {
+		tempScript, err := createTempScript(scriptConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		scriptConfig.Path = tempScript
+	}
+
+	if scriptConfig.Type == "" {
+		fileExtension := filepath.Ext(scriptConfig.Path)
+		switch fileExtension {
+		case ".sh":
+			scriptConfig.Type = ScriptTypeBash
+		case ".ps1":
+			scriptConfig.Type = ScriptTypePowershell
+		default:
+			return nil, fmt.Errorf(
+				"script with file extension '%s' is not valid. Only '.sh' and '.ps1' are supported",
+				fileExtension,
+			)
+		}
+	}
+
+	switch scriptConfig.Type {
+	case ScriptTypeBash:
+		return bash.NewBashScript(h.commandRunner, h.cwd, h.envVars), nil
+	case ScriptTypePowershell:
+		return powershell.NewPowershellScript(h.commandRunner, h.cwd, h.envVars), nil
+	default:
+		return nil, fmt.Errorf(
+			"script type '%s' is not a valid option. Only Bash and powershell scripts are supported",
+			scriptConfig.Type,
+		)
+	}
+}
+
+func createTempScript(scriptConfig *ScriptConfig) (string, error) {
+	scriptBytes := []byte(scriptConfig.Script)
+	hash := sha256.Sum256(scriptBytes)
+
+	var ext string
+	scriptHeader := []string{}
+
+	switch scriptConfig.Type {
+	case ScriptTypeBash:
+		ext = "sh"
+		scriptHeader = []string{
+			"#!/bin/bash",
+			"set -euo pipefail",
+		}
+	case ScriptTypePowershell:
+		ext = "ps1"
+	}
+
+	filename := fmt.Sprintf("%s.%s", base64.URLEncoding.
```

Comment:
How would you feel about instead just using `os.TempFile` to create a file with a random name in temp to use for the script contents?

Reply:
Updated to use `os.CreateTemp(...)` but still storing in `.azure/hooks`.  In a future iteration we may consider adding in convention based registration.  Imaging just creating a file called like `predeploy.sh` and just by it existing in the `.azure/hooks` folder it will be automatically registered without any azure.yaml registration requirement.